### ISSUE TITLE
error: Error loading target specification: Could not find specification for target "riscv64-unknown-linux-gnu".

### ISSUE THREAD

**Issue by andreas-schwab** [rust-lang/rust#123161]
https://build.opensuse.org/package/live_build_log/openSUSE:Factory:RISCV/flake-pilot/standard/riscv64

+ make compile_sci_static
cd firecracker-pilot/guestvm-tools/sci && RUSTFLAGS='-C target-feature=+crt-static' cargo build -v --profile static --target riscv64-unknown-linux-gnu
error: failed to run `rustc` to learn about target-specific information

Caused by:
  process didn't exit successfully: `rustc - --crate-name ___ --print=file-names -C target-feature=+crt-static --target riscv64-unknown-linux-gnu --crate-type bin --crate-type rlib --crate-type dylib --crate-type cdylib --crate-type staticlib --crate-type proc-macro --print=sysroot --print=split-debuginfo --print=crate-name --print=cfg` (exit status: 1)
  --- stderr
  error: Error loading target specification: Could not find specification for target "riscv64-unknown-linux-gnu". Run `rustc --print target-list` for a list of built-in targets

make: *** [Makefile:72: compile_sci_static] Error 101

**Comment by Nilstrieb** [comment]
This target does indeed not exist. Maybe you want `riscv64gc-unknown-linux-gnu`? As the error said, check `rustc --print target-list` for all available targets.

**Comment by andreas-schwab** [comment]
Why does it not exist?  That does not make any sense.

**Comment by bjorn3** [comment]
The base 64bit RISC-V ISA (RV64I) is not usable outside of microcontrollers. You need a bunch of extensions to make it usable. Linux mandates the G extension (which is an alias for the I, M, A, F, D, Zicsr and Zifencei extensions) and the C extension. Making the rust target arch riscv64gc. As such we only support riscv64gc-unknown-linux-gnu as Linux target and not riscv64-unknown-linux-gnu which would imply the lack of multiplication, atomics and floating point instructions as well as the lack of 16bit compressed instructions despite those being mandatory for Linux support. For bare metal targets we also have riscv64imac which doesn't allow floating point operations.

**Comment by andreas-schwab** [comment]
That doesn't make any sense.  This is a full featured target.

**Comment by bjorn3** [comment]
riscv64 without the G and C extensions is not a full featured target. Note that the riscv64-unknown-linux-gnu target in Clang corresponds to riscv64gc-unknown-linux-gnu in rustc as the G and C extensions are enabled by clang for riscv64-unknown-linux-gnu. Rustc target triples don't always exactly match the target triples that GCC and Clang use.

**Comment by andreas-schwab** [comment]
That does not make sense.  riscv64-linux-gnu is a fully featured target.
If rustc does not support that, it is terminally broken.

**Comment by bjorn3** [comment]
What you are calling riscv64-linux-gnu, rustc calls riscv64gc-unknown-linux-gnu.

**Comment by andreas-schwab** [comment]
That's why it is broken.  Everybody calls it riscv64-linux-gnu.

**Comment by workingjubilee** [comment]
@andreas-schwab You forgot the -unknown-.

**Comment by andreas-schwab** [comment]
It doesn't matter.  It's broken either way.

**Comment by workingjubilee** [comment]
@andreas-schwab 

And no, not everyone means that. When you pass `--target=riscv64gc-unknown-linux-gnu` to clang, it is in fact RISC-V 64-bit with Integer instructions only:

https://github.com/llvm/llvm-project/blob/84780af4b02cb3b86e4cb724f996bf8e02f2f2e7/llvm/docs/RISCVUsage.rst#base-isas

So in order to use common instructions like, for instance, multiplication and division, those must be specified with `-march`. If you are displeased with this knowledge, please go file a bug report with clang. And flake-pilot. And probably gcc as well.

**Comment by bjorn3** [comment]
> That's why it is broken. Everybody calls it riscv64-linux-gnu.

Not more broken than rustc's i686-unknown-linux-gnu being called i786-unknown-linux-gnu by GCC and Clang instead. (SSE2 support is mandated form rustc's i686-unknown-linux-gnu target)

We don't have a policy that mandates matching the target names of GCC and Clang. wasm32-wasip1 is still called wasm32-wasi by clang. (And clang additionally allows wasm32-unknown-wasi, just like it accepts x86_64-linux-gnu in the place of x86_64-unknown-linux-gnu.) And there are targets which aren't supported by upstream GCC and Clang aarch64-kmc-solid_asp3. If they add upstream support they may choose an entirely different target name, so in those cases there is nothing we could have done to prevent a divergence.

Also while we could have called the linux target riscv64-unknown-linux-gnu, for bare metal targets we have to use the suffixes for the enabled riscv extensions as unlike with GCC and Clang in rustc all code compiled for a single target must be ABI compatible with each other, but say riscv32i-unknown-none-elf and riscv32imafc-unknown-none-elf are not ABI compatible as the former passes floats in integer registers, while the latter passes them in float registers.

**Comment by andreas-schwab** [comment]
That doesn't make any sense.  Everbody calls the full featured target
riscv64-linux.  rustc must support it, or it will remain broken forever.

**Comment by workingjubilee** [comment]
I suppose we will remain broken forever, then!

**Comment by andreas-schwab** [comment]
It doesn't make any sense to force everyone add a special case just for
riscv.

**Comment by Nilstrieb** [comment]
don't worry, you have to add other special cases too like i686 or Wasi

**Comment by workingjubilee** [comment]
@andreas-schwab It doesn't make sense for clang to demand everyone write out `--target=riscv64-unknown-linux-gnu -march=m` just to get the version of RISCV64 that has multiplication instructions, yet here we are.

**Comment by andreas-schwab** [comment]
No, x86 works out of the box.

**Comment by workingjubilee** [comment]
Yes, compiling for a CISC chip with monotonically-increasing featuresets works differently than for a RISC chip which allows picking and choosing random featuresets in actual implementations. Why should the "out of the box" experience be the same for wildly different hardware?

**Comment by andreas-schwab** [comment]
x86 does not have monotonically increasing feature sets either.  Yet it is
not broken.

**Comment by Nilstrieb** [comment]
I don't think it's useful to have a discussion in that tone. rustc and clang simply diverge in their target naming, no one is inherently correct. I think rustc is more coherent with it's RISC-V target naming as described, but I definitely understand your frustration, inconsistencies are painful too.
Do note that this target is by far no the only target where rustc's name diverges from clang or gcc, there are others like Apple silicon being aarch64 vs arm64.
I don't think we're going to change the target name.

**Comment by andreas-schwab** [comment]
There is no need to change any existing targets, just support what
everyone else already uses.

**Comment by workingjubilee** [comment]
What would you prefer we do if "everyone else" disagrees, one day, on whether riscv64-unknown-linux-gnu means riscv64gc or, say, riscv64gcv_Zba_Zbb_Zbs? We do have a target where the latter is what "riscv64" means, already, and it even is a Linux target, though it is the one which uses the bionic libc.

**Comment by tgross35** [comment]
Think this issue can be closed, we won't be taking any action here. At this point a plain `riscv64` is ambiguous between `riscv64gc` and `riscv64a23`, possibly others in the future.

We could probably at least give a hint though, wrote that up in https://github.com/rust-lang/rust/issues/155085.
